src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/HMACHashProvider.Browser.Managed.cs

```diff
@@ -0,0 +1,140 @@
+                HashAlgorithmNames.SHA384 => (128, 384 / 8),
+                HashAlgorithmNames.SHA512 => (128, 512 / 8),
+                _ => throw new CryptographicException(SR.Format(SR.Cryptography_UnknownHashAlgorithm, hashAlgorithmId)),
+            };
+
+            _key = InitializeKey(key);
+            (_inner, _outer) = UpdateIOPadBuffers();
+        }
+
+        private byte[] InitializeKey(ReadOnlySpan<byte> key)
+        {
+            if (key.Length > _blockSizeValue)
+            {
+                byte[] result = new byte[_hashSizeValue];
+                _hash1.AppendHashData(key);
+                int written = _hash1.FinalizeHashAndReset(result);
+                Debug.Assert(written == result.Length);
+
+                return result;
+            }
+
+            return key.ToArray();
+        }
+
+        private (byte[] inner, byte[] outer) UpdateIOPadBuffers()
+        {
+            byte[] inner = new byte[_blockSizeValue];
+            byte[] outer = new byte[_blockSizeValue];
+
+            int i;
+            for (i = 0; i < _blockSizeValue; i++)
+            {
+                inner[i] = 0x36;
+                outer[i] = 0x5C;
+            }
+            for (i = 0; i < _key.Length; i++)
+            {
+                inner[i] ^= _key[i];
+                outer[i] ^= _key[i];
+            }
+
+            return (inner, outer);
+        }
+
+        public override void AppendHashData(ReadOnlySpan<byte> data)
+        {
+            if (_hashing == false)
+            {
+                _hash1.AppendHashData(_inner);
+                _hashing = true;
+            }
+
+            _hash1.AppendHashData(data);
+        }
+
+        public override int FinalizeHashAndReset(Span<byte> destination)
+        {
+            int written = GetCurrentHash(destination);
+            Reset();
+            return written;
+        }
+
+        public override int GetCurrentHash(Span<byte> destination)
+        {
+            if (_hashing == false)
+            {
+                _hash1.AppendHashData(_inner);
+                _hashing = true;
+            }
+
+            // finalize the original hash
+            Span<byte> hashValue1 = stackalloc byte[_hashSizeValue];
+            int hash1Written = _hash1.GetCurrentHash(hashValue1);
+            Debug.Assert(hash1Written == hashValue1.Length);
+
+            // write the outer array
+            _hash2.AppendHashData(_outer);
+            // write the inner hash and finalize the hash
+            _hash2.AppendHashData(hashValue1);
+            return _hash2.FinalizeHashAndReset(destination);
+        }
+
+        public override int HashSizeInBytes => _hashSizeValue;
+
+        public override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                _hash1.Dispose();
+                _hash2.Dispose();
+                Array.Clear(_inner);
+                Array.Clear(_outer);
+                Array.Clear(_key);
```

Comment:
In the world where we decide that the JIT is allowed to make the same optimizations that the C compiler does, and is allowed to skip writing to memory when it can "prove" that no one will read the value.

I don't expect it to happen, but it's just... one of those things.

Reply:
> but it's just... one of those things

...that's unnecessary? ;-)

Reply:
It strikes me as very counterintuitive that we care so much about clearing these arrays that we're going to call a method that avoids optimizations that won't ever be performed, but not enough to clear them if the object is finalized.

Reply:
> new up an HMACSHA256 with a key, do some stuff, dispose it, memory dump, and then complain if they see any residuals in the memory dump

These arrays aren't pinned.  Even with aggressively calling clear, there's of course no guarantee such a test would succeed.

Reply:
I don't know how much we want to stray from the original reference source implementation, but AFAICT the `inner` and `outer` arrays are really only used once (inner when the first write happens, outer when finalize happens). Those two arrays could just be `stackalloc`d and initialized when needed. That way they aren't put on the heap.

That doesn't solve the overall problem though, since the key would still need to be stored / cleared.

Reply:
> Those two arrays could just be stackallocd and initialized when needed. That way they aren't put on the heap.

Sounds like a reasonable change to me.

Reply:
The drawback to that approach is if someone holds onto a `HMACSHA256` instance and reuses it, we would be re-initializing this data each time the caller used it. Although the initialization isn't very intensive: filling the buffer with a constant and then XOR'ing with the key.